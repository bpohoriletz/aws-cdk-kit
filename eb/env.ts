import { Stack, CfnOutput } from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { CfnApplication, CfnEnvironment } from 'aws-cdk-lib/aws-elasticbeanstalk';
import * as iam from 'aws-cdk-lib/aws-iam';

import SecurityGroupDirector from '../ec2/directors/security-group';
import SecurityGroupBuilder from '../ec2/security-group-builder';
import { envOptionSettings } from './setting';
import * as con from '../utils/naming';

export async function createEnvironment(
  application: CfnApplication,
  resourceNamePrefix: string[],
  instanceProfile: iam.IInstanceProfile,
  vpc: ec2.IVpc,
  stack: Stack,
  solutionStackName: string
): Promise<[CfnEnvironment, ec2.SecurityGroup]> {
  const environmentName = con.ebEnvironmentName(resourceNamePrefix);
  const securityGroups = createSecurityGroups(stack, vpc);

  const env = new CfnEnvironment(stack, environmentName, {
    applicationName: application.applicationName!,
    environmentName: environmentName,
    solutionStackName: solutionStackName,
    optionSettings: await envOptionSettings(resourceNamePrefix, instanceProfile, securityGroups, vpc),
  });
  const cfnInstanceProfile = instanceProfile.node.defaultChild as iam.CfnInstanceProfile;
  env.addDependency(cfnInstanceProfile);
  env.addDependency(application);

  new CfnOutput(stack, `${environmentName}ApplicationUrl`, {
    exportName: `${environmentName.toLowerCase()}-application-url`,
    value: env.attrEndpointUrl,
  });

  return [env, securityGroups[1]];
}

// Private
function createSecurityGroups(stack: Stack, vpc: ec2.IVpc): ec2.SecurityGroup[] {
  const lbSecurityGroup = new SecurityGroupDirector(SecurityGroupBuilder).constructWebSecurityGroup(
    stack,
    'LbSecurityGroup',
    vpc
  );
  const webSecurityGroup = new SecurityGroupDirector(SecurityGroupBuilder).constructWebSecurityGroup(
    stack,
    'WebSecurityGroup',
    vpc
  );
  webSecurityGroup.connections.allowFrom(new ec2.Connections({ securityGroups: [lbSecurityGroup] }), ec2.Port.tcp(80));

  return [lbSecurityGroup, webSecurityGroup];
}
