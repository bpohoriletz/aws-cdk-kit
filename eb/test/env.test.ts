import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { stub } from '../../test/stubs';

import * as func from '../env';

describe('ElasticBeanstalk .createEnvironment()', () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test('matches snapshot', async () => {
    await func.createEnvironment(
      stub(stack, 'eb.CfnApplication'),
      ['pre', 'fix'],
      stub(stack, 'iam.InstanceProfile'),
      stub(stack, 'ec2.Vpc'),
      stack,
      'any'
    );

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
