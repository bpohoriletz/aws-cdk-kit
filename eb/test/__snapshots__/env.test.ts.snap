// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticBeanstalk .createEnvironment() matches snapshot 1`] = `
{
  "Outputs": {
    "prefixenvApplicationUrl": {
      "Export": {
        "Name": "pre-fix-env-application-url",
      },
      "Value": {
        "Fn::GetAtt": [
          "prefixenv",
          "EndpointURL",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CfnApplicationID": {
      "Properties": {
        "ApplicationName": "eb.CfnApplication",
      },
      "Type": "AWS::ElasticBeanstalk::Application",
    },
    "InstanceProfileIDE79A37AF": {
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceProfileIDInstanceRole9444F891",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "InstanceProfileIDInstanceRole9444F891": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LbSecurityGroupEF17985E": {
      "Properties": {
        "GroupDescription": "Web security group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcID703024FC": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIDIGW70D85C75": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcIDIsolatedSubnet1RouteTableAssociationC2ECBCF5": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDIsolatedSubnet1RouteTableCF9B88CE",
        },
        "SubnetId": {
          "Ref": "VpcIDIsolatedSubnet1SubnetFCDAB651",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDIsolatedSubnet1RouteTableCF9B88CE": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDIsolatedSubnet1SubnetFCDAB651": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/IsolatedSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDIsolatedSubnet2RouteTable9A3A82EF": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDIsolatedSubnet2RouteTableAssociation88968DE6": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDIsolatedSubnet2RouteTable9A3A82EF",
        },
        "SubnetId": {
          "Ref": "VpcIDIsolatedSubnet2Subnet52CFD445",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDIsolatedSubnet2Subnet52CFD445": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.96.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/IsolatedSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDPrivateSubnet1DefaultRouteD7C53805": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "VpcIDPublicSubnet1NATGateway3C86780D",
        },
        "RouteTableId": {
          "Ref": "VpcIDPrivateSubnet1RouteTableFC20DF64",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcIDPrivateSubnet1RouteTableAssociation43036593": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDPrivateSubnet1RouteTableFC20DF64",
        },
        "SubnetId": {
          "Ref": "VpcIDPrivateSubnet1SubnetE31736A9",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDPrivateSubnet1RouteTableFC20DF64": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDPrivateSubnet1SubnetE31736A9": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDPrivateSubnet2DefaultRoute21C5563B": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "VpcIDPublicSubnet2NATGatewayF1C39B22",
        },
        "RouteTableId": {
          "Ref": "VpcIDPrivateSubnet2RouteTable9348734A",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcIDPrivateSubnet2RouteTable9348734A": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDPrivateSubnet2RouteTableAssociationD9C8D78D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDPrivateSubnet2RouteTable9348734A",
        },
        "SubnetId": {
          "Ref": "VpcIDPrivateSubnet2SubnetC6B60E3F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDPrivateSubnet2SubnetC6B60E3F": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.160.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDPublicSubnet1DefaultRoute73EE5304": {
      "DependsOn": [
        "VpcIDVPCGWE6968D99",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "VpcIDIGW70D85C75",
        },
        "RouteTableId": {
          "Ref": "VpcIDPublicSubnet1RouteTableA9496B93",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcIDPublicSubnet1EIP66CFB463": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcIDPublicSubnet1NATGateway3C86780D": {
      "DependsOn": [
        "VpcIDPublicSubnet1DefaultRoute73EE5304",
        "VpcIDPublicSubnet1RouteTableAssociation2D570C68",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "VpcIDPublicSubnet1EIP66CFB463",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "VpcIDPublicSubnet1Subnet49ECF0B5",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcIDPublicSubnet1RouteTableA9496B93": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDPublicSubnet1RouteTableAssociation2D570C68": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDPublicSubnet1RouteTableA9496B93",
        },
        "SubnetId": {
          "Ref": "VpcIDPublicSubnet1Subnet49ECF0B5",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDPublicSubnet1Subnet49ECF0B5": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDPublicSubnet2DefaultRoute16A93887": {
      "DependsOn": [
        "VpcIDVPCGWE6968D99",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "VpcIDIGW70D85C75",
        },
        "RouteTableId": {
          "Ref": "VpcIDPublicSubnet2RouteTable7E67B44D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcIDPublicSubnet2EIPF974E61A": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "VpcIDPublicSubnet2NATGatewayF1C39B22": {
      "DependsOn": [
        "VpcIDPublicSubnet2DefaultRoute16A93887",
        "VpcIDPublicSubnet2RouteTableAssociation5A90573C",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "VpcIDPublicSubnet2EIPF974E61A",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "VpcIDPublicSubnet2SubnetBF551D4B",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VpcIDPublicSubnet2RouteTable7E67B44D": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcIDPublicSubnet2RouteTableAssociation5A90573C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "VpcIDPublicSubnet2RouteTable7E67B44D",
        },
        "SubnetId": {
          "Ref": "VpcIDPublicSubnet2SubnetBF551D4B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcIDPublicSubnet2SubnetBF551D4B": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/VpcID/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcIDVPCGWE6968D99": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "VpcIDIGW70D85C75",
        },
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "WebSecurityGroup73AF7387": {
      "Properties": {
        "GroupDescription": "Web security group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": {
          "Ref": "VpcID703024FC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "WebSecurityGroupfromLbSecurityGroup80C0FEE8CF": {
      "Properties": {
        "Description": "from LbSecurityGroup:80",
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "WebSecurityGroup73AF7387",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LbSecurityGroupEF17985E",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "prefixenv": {
      "DependsOn": [
        "CfnApplicationID",
        "InstanceProfileIDE79A37AF",
      ],
      "Properties": {
        "ApplicationName": "eb.CfnApplication",
        "EnvironmentName": "pre-fix-env",
        "OptionSettings": [
          {
            "Namespace": "aws:autoscaling:launchconfiguration",
            "OptionName": "SecurityGroups",
            "Value": {
              "Fn::GetAtt": [
                "WebSecurityGroup73AF7387",
                "GroupId",
              ],
            },
          },
          {
            "Namespace": "aws:elbv2:loadbalancer",
            "OptionName": "SecurityGroups",
            "Value": {
              "Fn::GetAtt": [
                "LbSecurityGroupEF17985E",
                "GroupId",
              ],
            },
          },
          {
            "Namespace": "aws:ec2:vpc",
            "OptionName": "ELBSubnets",
            "Value": {
              "Fn::Join": [
                "",
                [
                  {
                    "Ref": "VpcIDPublicSubnet1Subnet49ECF0B5",
                  },
                  ",",
                  {
                    "Ref": "VpcIDPublicSubnet2SubnetBF551D4B",
                  },
                ],
              ],
            },
          },
          {
            "Namespace": "aws:ec2:vpc",
            "OptionName": "Subnets",
            "Value": {
              "Fn::Join": [
                "",
                [
                  {
                    "Ref": "VpcIDPrivateSubnet1SubnetE31736A9",
                  },
                  ",",
                  {
                    "Ref": "VpcIDPrivateSubnet2SubnetC6B60E3F",
                  },
                ],
              ],
            },
          },
          {
            "Namespace": "aws:autoscaling:launchconfiguration",
            "OptionName": "IamInstanceProfile",
            "Value": {
              "Ref": "InstanceProfileIDE79A37AF",
            },
          },
        ],
        "SolutionStackName": "any",
      },
      "Type": "AWS::ElasticBeanstalk::Environment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
